### PR TITLE
3D Tiles Styling - Check for type mismatches in JS and shader backends

### PR DESCRIPTION
**Pull request by lilleyse**
Based on discussion here: https://github.com/AnalyticalGraphicsInc/3d-tiles/pull/166

To start this goes through the unary and binary operators and throws errors if the arguments are not the same type. There's a bit more work to be done with the math library, vector constructors, etc.

In some ways this PR is starting the process of changing the styling language to not be as dependent on JavaScript type-conversion rules. Is the end goal of the spec to not rely on JavaScript conventions at all?

### CODE REVIEW

**Comment by lilleyse** [comment]
Some tests are broken and I need to write new tests. I don't want to get too far into this unless this is the right direction.

File: Source/Scene/Expression.js [hunk 1]
```diff
@@ -1060,8 +1104,17 @@ define([
             return Cartesian3.add(left, right, ScratchStorage.getCartesian3());
         } else if ((right instanceof Cartesian4) && (left instanceof Cartesian4)) {
             return Cartesian4.add(left, right, ScratchStorage.getCartesian4());
+        } else if ((typeof(left) === 'string') || (typeof(right) === 'string')) {
+            return left + right;
```
Comment (lilleyse) [inline Source/Scene/Expression.js]: This relies on each types `toString` function. Should the spec be explicit about how each type is stringified?

File: Source/Scene/Expression.js [hunk 2]
```diff
@@ -960,7 +960,13 @@ define([
     // that we can assign if we know the types before runtime
 
     Node.prototype._evaluateNot = function(frameState, feature) {
-        return !(this._left.evaluate(frameState, feature));
+        var left = this._left.evaluate(frameState, feature);
+        //>>includeStart('debug', pragmas.debug);
+        if (typeof(left) !== 'boolean') {
+            throw new DeveloperError('Error: Operation is undefined.');
```
Comment (pjcozzi) [inline Source/Scene/Expression.js]: These exceptions are all fine, but can we be more precise about the type mismatch and any other info we can provide to help devs track down the issue?

Comment (mramato) [inline Source/Scene/Expression.js]: There's also need need to include `Error: ` in the message, you are throwing an error so it's redundant.

File: Source/Scene/Expression.js (hunk 1 above)
Comment (pjcozzi) [inline Source/Scene/Expression.js]: See what JavaScript says about `toString` functions.  I want to say yes, we should state how, so things are very well specified, but we would also have to think about how many decimal places, etc.

**Comment by pjcozzi** [comment]
> I don't want to get too far into this unless this is the right direction.

Looks reasonable to me.  Do you have any concerns?

**Comment by pjcozzi** [comment]
@lilleyse please bump when ready.

**Comment by lilleyse** [comment]
Ready now.

@Dylan-Brown you'll want to do similar type checking for the other built-in functions in https://github.com/AnalyticalGraphicsInc/cesium/issues/4865

**Comment by pjcozzi** [comment]
Looks good!

**Pull request merged**